### PR TITLE
Fix significant performance issue of fixed-step stochastic jump solver

### PR DESCRIPTION
**Pull request by pierreguilmin**
This fixes a major performance issue with our fixed-step jump solvers. On CPU, the replacement of `jax.lax.cond` by `jax.lax.select` (see e.g. https://github.com/jax-ml/jax/issues/7934) leads to a x100 speedup for trajectories simulation of a qubit.

| Number of trajectories | Duration before the PR         | Duration after the PR | Speedup |
|---------------|---------------------|----------------|-|
| 1,000    |      12.7 s ± 1.36 s  | 119 ms ± 8.87 ms    | x106            |
| 10,000           | 158 s (1 run)      |       1.1 s ± 186 ms          |  x144 |
| 100,000           | ??     | 13.1 s ± 1.54 s per loop        | ??|

The new perfs on GPU are quite impressive:
| Number of trajectories | Duration after the PR (GPU) |
|---------------|---------------------|
| 1,000 | 27.5 ms ± 258 μs per loop |
| 10,000 | 50 ms ± 137 μs per loop |
| 100,000 | 185 ms ± 609 μs per loop |

<details>
  <summary>Benchmark code</summary>

```python
import dynamiqs as dq
import numpy as np
import jax

H = dq.zeros(2)
jump_ops = [dq.sigmam(), dq.sigmap()]
psi0 = dq.excited()
tsave = np.linspace(0, 2.0, 101)
key = jax.random.PRNGKey(42)
method = dq.method.EulerJump(dt=1e-3)
for ntrajs in [int(1e3), int(1e4)]:
    keys = jax.random.split(key, ntrajs)

    def foo():
        return dq.jssesolve(
            H, jump_ops, psi0, tsave, keys, method=method
        ).states.block_until_ready()

    %timeit -r1 -n1 -q foo()
    %timeit foo()
```

</details>

There is another speedup to implement for `jssesolve` vs `jsmesolve`, the same as #911. I can do it once this PR is merged.

### CODE REVIEW

File: dynamiqs/integrators/core/fixed_step_stochastic_integrator.py [hunk 1]
```diff
@@ -214,21 +207,36 @@ class JumpSolveFixedStepIntegrator(StochasticSolveFixedStepIntegrator):
...
         saved = super().postprocess_saved(saved, ylast.state)
-        return JumpSolveSaved(saved.ysave, saved.extra, saved.Esave, ylast.clicktimes)
+
+        # convert array of click value at each time to array of clicktimes, for example:
+        #   times = [0, 10, 20, 30, 40, 50]
+        #   clicks = [0, 1, 0, 1, 0, 2]
+        #   => clicktimes = [[10, 30, nan, ...], [50, nan, nan, ...]]
+        clicktimes = jnp.full((self.nmeas, self.options.nmaxclick), jnp.nan)
+        for jump_idx in jnp.arange(self.nmeas):
+            times = jnp.arange(self.t0, self.t1, self.dt)
```
Comment (gautierronan) [inline dynamiqs/integrators/core/fixed_step_stochastic_integrator.py]: There's sometimes weird stuff with the length of `jnp.arange` when `(self.t1 - self.t0) / self.dt` is exactly an integer. Might be safer to write `times = self.t0 + jnp.linspace(nsteps) * self.dt` here, to make sure the length is always `nsteps`.

File: dynamiqs/integrators/core/fixed_step_stochastic_integrator.py [hunk 2]
```diff
@@ -214,21 +207,36 @@ class JumpSolveFixedStepIntegrator(StochasticSolveFixedStepIntegrator):
...
         return super().save(y.state)
 
     def postprocess_saved(self, saved: Saved, ylast: PyTree) -> Saved:
         saved = super().postprocess_saved(saved, ylast.state)
-        return JumpSolveSaved(saved.ysave, saved.extra, saved.Esave, ylast.clicktimes)
+
+        # convert array of click value at each time to array of clicktimes, for example:
+        #   times = [0, 10, 20, 30, 40, 50]
+        #   clicks = [0, 1, 0, 1, 0, 2]
+        #   => clicktimes = [[10, 30, nan, ...], [50, nan, nan, ...]]
```
Comment (gautierronan) [inline dynamiqs/integrators/core/fixed_step_stochastic_integrator.py:235]: I don't quite understand this: how can `clicks[-1] == 2`? Shouldn't it always be 0 or 1? And how do you know which jump operator the click is associated to?

**Review by gautierronan — APPROVED** [review]
That's a crazy speedup indeed. Definitely something to be very careful about when using jax.lax.cond!

Two comments below, but mostly related to the fact that I don't understand the code of this fixed step stochastic class very well. Not blocking merging if you think they're not relevant.

File: dynamiqs/integrators/core/fixed_step_stochastic_integrator.py (hunk 2 above)
Comment (pierreguilmin) [reply dynamiqs/integrators/core/fixed_step_stochastic_integrator.py:235]: For this specific implementation, it's easier to have a single `clicks` array which labels the jump channel that clicked (0 = no click, 1 = first jump operator detector clicked, 2 = second jump operator detector clicked, etc.). We convert it back to `clicktimes` in the end.
